File: Solution/3152.cpp
class Solution {
public:
    vector<bool> isArraySpecial(vector<int>& nums, vector<vector<int>>& queries) {
        const int n = nums.size();
        const int m = queries.size();
        vector<int> mem(n, 0);

        for (int i = 1; i < n; ++i) {
            if (nums[i-1] % 2 == nums[i] % 2) mem[i] = mem[i-1] + 1;
            else mem[i] = mem[i-1];
        }

        vector<bool> ans(m, true);
        for (int i = 0; i < m; ++i) {
            int from = queries[i][0], to = queries[i][1];
            if (mem[from] != mem[to]) ans[i] = false; 
        }
        return ans;
    }
};

File: Solution/3152.py
class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        prefix_sum = [0]
        for i in range(len(nums)):
            bad = 1
            if i == 0 or (nums[i] % 2 != nums[i-1] % 2) :
                bad = 0
            prefix_sum.append(prefix_sum[-1] + bad)
        
        print(prefix_sum)
        out = []
        for query in queries:
            s, e = query
            range_sum = prefix_sum[e+1] - prefix_sum[s+1]
            if range_sum == 0:
                out.append(True)
            else:
                out.append(False)
        return out